File: server/apps/nexotype/schemas/clinical/biomarker_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class BiomarkerCreate(BaseModel):
    """Schema for creating a new biomarker"""
    name: str
    loinc_code: str | None = None


class BiomarkerUpdate(BaseModel):
    """Schema for updating a biomarker"""
    name: str = None
    loinc_code: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class BiomarkerDetail(BaseModel):
    """Schema for biomarker details"""
    id: int
    name: str
    loinc_code: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class BiomarkerListResponse(BaseModel):
    """Response schema for listing biomarkers"""
    success: bool
    data: list[BiomarkerDetail] | None = None
    count: int = 0
    error: str | None = None


class BiomarkerResponse(BaseModel):
    """Response schema for single biomarker operations"""
    success: bool
    data: BiomarkerDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/clinical/biomarker.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const BiomarkerSchema = z.object({
  id: z.number(),
  name: z.string().min(1).max(255),
  loinc_code: z.string().max(20).nullable().optional(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateBiomarkerSchema = z.object({
  name: z.string().min(1).max(255),
  loinc_code: z.string().max(20).nullable().optional(),
});

export const UpdateBiomarkerSchema = z.object({
  name: z.string().min(1).max(255).optional(),
  loinc_code: z.string().max(20).nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type Biomarker = z.infer<typeof BiomarkerSchema>;
export type CreateBiomarker = z.infer<typeof CreateBiomarkerSchema>;
export type UpdateBiomarker = z.infer<typeof UpdateBiomarkerSchema>;

// ==========================================
// Response Types
// ==========================================

export type BiomarkerResponse = {
  success: boolean;
  data?: Biomarker;
  error?: string;
};

export type BiomarkersResponse = {
  success: boolean;
  data?: Biomarker[];
  count?: number;
  error?: string;
};
